sympy/series/sequences.py

```diff
@@ -0,0 +1,522 @@
+        else:
+            initial = self.start
+
+        if step == None:
+            if self.start is S.NegativeInfinity:
+                step = -self.step
+            else:
+                step = self.step
+
+        return initial + i*step
+
+    def __getitem__(self, index):
+        if isinstance(index, integer_types):
+            index = self._ith_point(index)
+            return self.coeff(index)
+        elif isinstance(index, slice):
+            start, stop = index.start, index.stop
+            if start == None:
+                start = 0
+            if stop == None:
+                stop = self.length
+            return [self.coeff(self._ith_point(i, index.step)) for i in\
+                               range(start, stop)]
+
+
+class EmptySequence(with_metaclass(Singleton, SeqBase)):
+    """
+    Represents an empty sequence. The empty sequence is available as a
+    singleton as S.EmptySequence.
+
+    Examples
+    ========
+
+    >>> from sympy import S
+    >>> S.EmptySequence
+    EmptySequence()
+    """
+
+    is_iterable = True
+
+    @property
+    def _interval(self):
+        return S.EmptySet
+
+    @property
+    def _length(self):
+        return S.Zero
+
+    def __iter__(self):
+        return iter([])
+
+
+def _parse_interval(interval):
+    """
+    interval should be of the form (start, step) or (start, stop, step)
+    Both start and stop cannot be unbounded
+    step cannot be unbounded
+
+    Allowed:
+    * Any instance of set
+    * (Set, step)
+    * (start, stop)
+    * (start, stop, step)
+
+    returns an Interval object and a step value
+
+    Examples
+    ========
+
+    >>> from sympy.series.sequences import _parse_interval as pari
+    >>> from sympy import Interval
+    >>> pari(Interval(0, 5))
+    ([0, 5], 1)
+    >>> pari((Interval(0, 5), 2))
+    ([0, 5], 2)
+    >>> pari((0, 5))
+    ([0, 5], 1)
+    >>> pari((0, 5, 2))
+    ([0, 5], 2)
+    """
+    start, stop, step = None, None, None
+    if isinstance(interval, Set):
+        start, stop = interval.inf, interval.sup
+    elif is_sequence(interval, Tuple):
+        if len(interval) == 2:
+            if isinstance(interval[0], Set):
+                start, stop = interval[0].inf, interval[0].sup
+                step = interval[1]
+            else:
+                start, stop = interval
+        elif len(interval) == 3:
+            start, stop, step = interval
+
+    if step == None:
+        step = 1 # default
+
+    if start == None or stop == None:
+        raise ValueError('Invalid limits given: %s' % str(interval))
+
+    if start is S.NegativeInfinity and stop is S.Infinity:
+            raise ValueError(filldedent("""Both the start and end value\
+                                        cannot be unbounded"""))
+
+    if step in [S.NegativeInfinity, S.Infinity]:
+        raise ValueError("step cannot be unbounded")
+
+    return (Interval(start, stop), sympify(step))
+
+
+class SeqExpr(SeqBase):
```

Comment:
I'm still not sure about your class hierarchy. If you plan on implementing `SeqAdd`, `SeqMul`, or other such classes, then maybe a base class like this makes sense. Otherwise, `SeqBase` can probably just hold everything.